Expand conceptual docs around `ActionListener` (#105405)
elastic/elasticsearch: docs/internal/DistributedArchitectureGuide.md

### ActionListener

`ActionListener` is generally used as a continuation, in the sense of [_continuation-passing
style_](https://en.wikipedia.org/wiki/Continuation-passing_style) (CPS): an extra argument to a function which defines how to continue the
computation when the result is available. This is in contrast to _direct style_ which is the more usual style of calling methods that return
values directly back to the caller so they can continue executing as normal. There's essentially two ways that computation can continue in
Java (it can return a value or it can throw an exception) which is why `ActionListener` has both an `onResponse()` and an `onFailure()`

> method.

Reviewer: I think it is an improvement over the existing text. But `CPS` seems kind of exotic! Is there a reason we are not explicitly using the terms synchronous/asynchronous or maybe blocking/non-blocking call, or say `ActionListener` is a callback that can be passed to an async method, or something along those lines? Those seem to be more wide-spread, I feel. Could link to https://en.wikipedia.org/wiki/Asynchronous_method_invocation, for example.

Reviewer: (the source might not be the best, but maybe an option if we're looking for something similar to the CSP link)

Author: I think we're positively not doing AMI - that means turning a method `T X()` into two methods `AsyncResult BeginX()` and `T EndX(AsyncResult)` where you have to check `AsyncResult#isDone` every now and then and only call `EndX()` once the result is ready. Bleurgh.

I'm curious tho, why do you think CPS is exotic? It's exactly what we're doing. We have to do the CPS translation by hand because nobody has taught Java how to do it for us yet, but mostly when I'm reading code I'm translating it back into direct style in my head.

I did mention blocking a bit here, I'll see if I can get some mention of sync/async in too. They're somewhat orthogonal tho, lots of paths through our code will complete the listener before returning to the caller which means it's at least a little bit synchronous, and async code can involve blocking too (e.g. if it does disk IO).

Reviewer: > I think we're positively not doing AMI

Ignore the link. Bad suggestion! :)

> why do you think CPS is exotic? It's exactly what we're doing.

I'm not disagreeing with it, it just doesn't tell me much right away if I don't know the term. I think rather than defining the style, just simplifying that an `ActionListener` is a callback that can be passed around and completed synchronously, or stored and called back later from another thread seems enough, if that is not too much of a simplification. I'd rather see the simplified explanation that might gradually builds up to CPS maybe.  

All I'm saying is that if I can't grok the concept of action listeners from the code, another new term and going through the CPS generic definition to arrive at callback might be overdoing it. Is callback too inaccurate? I feel like that's more widely-known than CPS. 

I agree that  blocking/unblocking might not be the best way of formulating it.

Just an opinion, curious what others think.

Author: WDYT about 43dbbed07b8ad2b868c2f8793c897c05a1225aa4?

Reviewer: Great! Thanks for reworking it. :)

One small question: Would it fit right after the first paragraph (or at the end of it) to say: if a method receives an action listener, it should not return or throw, but do one of them ONLY via the action listener. Or is that generalization a bit unnecessary? I kind of think of like that most often, but I understand that we don't have a way of enforcing it, and there might be many counter-examples in the code-base.

Author: Yep, see e97eb4c3265715a74b8f8ffa4e4e1c4dc3df78bb.

### ActionListener

Callbacks are used extensively throughout Elasticsearch because they enable us to write asynchronous and nonblocking code, i.e. code which

> doesn't necessarily compute a result straight away but also doesn't block the calling thread waiting for the result to become available.

Author: Tried clarifying this within the sentence but it got so big that it fell apart. Expanded this out into a list instead in d8e3c31707c.

> Callbacks can be passed into methods for the method to perhaps complete straight away, or instead perhaps to store in a data structure and

Author: Reworded in the list of usages added in d8e3c31707c.

complete at some later time, or (most commonly) pass along down into some other method which takes a callback argument, possibly after
modifying its behaviour slightly by wrapping it in another callback.

> `ActionListener` is a general-purpose callback interface that is used extensively across the Elasticsearch codebase. It's an important

Author: Good point, gone in d8e3c31707c.

feature of the Elasticsearch codebase that `ActionListener` is used pretty much everywhere that needs to perform some asynchronous and
nonblocking computation. The uniformity makes it easier to compose parts of the system together without needing to build adapters to convert

> back and forth between different kinds of callback. It also makes it easier to develop the skills needed to read and understand all the

Author: Eh I feel this is useful to address this sticking point for new folks: this stuff is not obvious, it takes practice and conscious effort. I can drop it if you're sure but haven't done yet.

asynchronous code, although this definitely takes practice and is certainly not easy in an absolute sense. Finally, it has allowed us to
build a rich library for working with `ActionListener` instances themselves, creating new instances out of existing ones and completing them
in interesting ways. See for instance:

- all the static methods on [ActionListener](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java) itself
- [`ThreadedActionListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java) for forking work elsewhere
- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java) for running work in parallel
- [`SubscribableListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java) for constructing flexible workflows

> 

Reviewer 2: Since there's quite a few links now throughout, you could also use a markdown feature to defer the URLs. Definitely not necessary, but I don't know if you know about it.

Put here:
[ActionListener][]

[SubscribableListener][]

Add at the end of the section (it won't be visible if you separate it by a line from other text)
[ActionListener]: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java
[SubscribableListener]: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

Author: Yeah, I find that makes it harder to read the plaintext (and makes no difference to the rendered copy), although neither is particularly usable. The links to code throughout are very much indicating to me that this information should be in the `ActionListener` Javadocs which has proper support for code links.

Callback-based asynchronous code can easily call regular synchronous code, but synchronous code cannot run callback-based asynchronous code

> without blocking the calling thread until the continuation is executed. This blocking is at best undesirable (threads are too expensive to

Reviewer 2: "without blocking...."
=> 
something like,
"without blocking until the callback is executed (`onResponse()` or `onFailure()`)"

'calling thread' is confusing since the subject is already 'synchronous code'.
'Continuation' is the fancy term, but you've been using 'callback' until now, so it's confusing to suddenly introduce it mid-paragraph.

Author: Ah oops I missed that one. Fixed in d8e3c31707c.

waste with unnecessary blocking) and at worst outright broken (the blocking can lead to deadlock). Unfortunately this means that most of our
code ends up having to be written with callbacks, simply because it's ultimately calling into some other code that takes a callback. The

> entry points for all Elasticsearch APIs are callback-based, and the whole system works in terms of an event loop which processes network

Reviewer 2: "The entry points for all Elasticsearch APIs are callback-based"

Similar to my subsequent comment, maybe add a demonstrative code link, if you can think of one.

Author: Added links in d8e3c31707c.

> events via callbacks.

Author: Netty really. Added a link in d8e3c31707c.

`ActionListener` is not an _ad-hoc_ invention. Formally speaking, it is our implementation of the general concept of a continuation in the
sense of [_continuation-passing style_](https://en.wikipedia.org/wiki/Continuation-passing_style) (CPS): an extra argument to a function
which defines how to continue the computation when the result is available. This is in contrast to _direct style_ which is the more usual
style of calling methods that return values directly back to the caller so they can continue executing as normal. There's essentially two
ways that computation can continue in Java (it can return a value or it can throw an exception) which is why `ActionListener` has both an
`onResponse()` and an `onFailure()` method.

CPS is strictly more expressive than direct style: direct code can be mechanically translated into continuation-passing style, but CPS also
enables all sorts of other useful control structures such as forking work onto separate threads, possibly to be executed in parallel,
perhaps even across multiple nodes, or possibly collecting a list of continuations all waiting for the same condition to be satisfied before

> proceeding. Some languages have first-class support for continuations (e.g. the `async` and `await` primitives in C#) allowing the

Author: Linked to `SubscribableListener` but I don't think that's the only example.

programmer to write code in direct style away from those exotic control structures, but Java does not. That's why we have to manipulate all
the callbacks ourselves.

Strictly speaking, CPS requires that a computation _only_ continues by calling the continuation. In Elasticsearch, this means that

> asynchronous methods must have `void` return type and may not throw any exceptions. This is mostly the case in our code as written today,

Reviewer 2: maybe skip mention of void, to focus on exceptions, the important bit for this paragraph? Otherwise with the next sentence I get thinking that there are deviations from a void return type 😅

Author: There are deviations from a void return type too, because of Reasons™. E.g. `org.elasticsearch.index.seqno.ReplicationTracker#addRetentionLease` and other methods that synchronously update the retention leases locally and then asynchronously push the change out to the rest of the replication group.

and is a good guiding principle, but we don't enforce it and there are some deviations from this rule. In particular, it's not uncommon to

> permit some methods to throw an exception, using things like `ActionListener#run` (or an equivalent `try ... catch ...` block) further up

Author: ++ more reasons to use Javadoc for this.

the stack to handle it.

This pattern is often used in the transport action layer with the use of the

> [ChannelActionListener](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java)

Author: Yeah there's no good answer here apart from using Javadoc to avoid broken links, or else relying on this stuff hardly changing in `main` (and accepting that if it does we can still work out what's going on using the Git history).

### ActionListener

Callbacks are used extensively throughout Elasticsearch because they enable us to write asynchronous and nonblocking code, i.e. code which
doesn't necessarily compute a result straight away but also doesn't block the calling thread waiting for the result to become available.
They support several useful control flows:

- They can be completed immediately on the calling thread.
- They can be completed concurrently on a different thread.
- They can be stored in a data structure and completed later on when the system reaches a particular state.
- Most commonly, they can be passed on to other methods that themselves require a callback.

> - They can be wrapped in another callback which modifies the behaviour of the original callback, perhaps adding some extra code to run

Reviewer 2: I think saying using 'modifies' is misleading, as it is not possible to modify the original callback, only add logic to run before or after the original callback via a new callback.

> - They can be wrapped in another callback which modifies the behaviour of the original callback, perhaps adding some extra code to run

Author: Sure, but the new callback does not typically have entirely new _behaviour_, and that's what we're saying we're modifying. Adding something which runs before or after the original code is definitely a good way to modify its behaviour IMO.

- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java) for running work in parallel
- [`SubscribableListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java) for constructing flexible workflows

Callback-based asynchronous code can easily call regular synchronous code, but synchronous code cannot run callback-based asynchronous code
without blocking the calling thread until the callback is called back. This blocking is at best undesirable (threads are too expensive to
waste with unnecessary blocking) and at worst outright broken (the blocking can lead to deadlock). Unfortunately this means that most of our
code ends up having to be written with callbacks, simply because it's ultimately calling into some other code that takes a callback. The
entry points for all Elasticsearch APIs are callback-based (e.g. REST APIs all start at

> [`org.elasticsearch.rest.BaseRestHandler#prepareRequest`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java),

Reviewer 2: Just adding some line numbers to reference the method rather than the file.

and transport APIs all start at

> [`org.elasticsearch.action.support.TransportAction#doExecute`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/TransportAction.java))

Reviewer 2: ```suggestion
[`org.elasticsearch.action.support.TransportAction#doExecute`](https://github.com/elastic/elasticsearch/blob/v8.12.2/server/src/main/java/org/elasticsearch/action/support/TransportAction.java#L65))
```

CPS is strictly more expressive than direct style: direct code can be mechanically translated into continuation-passing style, but CPS also
enables all sorts of other useful control structures such as forking work onto separate threads, possibly to be executed in parallel,
perhaps even across multiple nodes, or possibly collecting a list of continuations all waiting for the same condition to be satisfied before
proceeding (e.g.
[`SubscribableListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java)
amongst many others). Some languages have first-class support for continuations (e.g. the `async` and `await` primitives in C#) allowing the
programmer to write code in direct style away from those exotic control structures, but Java does not. That's why we have to manipulate all
the callbacks ourselves.

Strictly speaking, CPS requires that a computation _only_ continues by calling the continuation. In Elasticsearch, this means that
asynchronous methods must have `void` return type and may not throw any exceptions. This is mostly the case in our code as written today,

> and is a good guiding principle, but we don't enforce it and there are some deviations from this rule. In particular, it's not uncommon to

Reviewer 2: ```suggestion
and is a good guiding principle, but we don't enforce void exceptionless methods and there are some deviations from this rule. In particular, it's not uncommon to
```

permit some methods to throw an exception, using things like

> [`ActionListener#run`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java)

Reviewer 2: ```suggestion
[`ActionListener#run`](https://github.com/elastic/elasticsearch/blob/v8.12.2/server/src/main/java/org/elasticsearch/action/ActionListener.java#L381-L390)
```

> (or an equivalent `try ... catch ...` block) further up the stack to handle it.

Reviewer 2: ```suggestion
(or an equivalent `try ... catch ...` block) further up the stack to handle it. Some methods also take (and may execute) an ActionListener parameter, but still return a value separately for other local synchronous work.
```
